app/src/handler.rs

```diff
@@ -0,0 +1,115 @@
+
+#[repr(u8)]
+pub enum Ins {
+    GetVersion,
+    GetPubKey,
+    SignHash,
+    SignTx,
+}
+
+impl TryFrom<io::ApduHeader> for Ins {
+    type Error = ErrorCode;
+    fn try_from(header: io::ApduHeader) -> Result<Self, Self::Error> {
+        match header.ins {
+            0 => Ok(Ins::GetVersion),
+            1 => Ok(Ins::GetPubKey),
+            2 => Ok(Ins::SignHash),
+            3 => Ok(Ins::SignTx),
+            _ => Err(ErrorCode::BadIns),
+        }
+    }
+}
+
+pub fn handle_apdu(
+    comm: &mut io::Comm,
+    ins: Ins,
+    sign_tx_context: &mut SignTxContext,
+    tx_reviewer: &mut TxReviewer,
+) -> Result<(), io::Reply> {
+    if comm.rx == 0 {
+        return Err(ErrorCode::BadLen.into());
+    }
+
+    let mut path: [u32; 5] = [0; 5];
+    let apdu_header = comm.get_apdu_metadata();
+    if apdu_header.cla != 0x80 {
+        return Err(ErrorCode::BadCla.into());
+    }
+
+    match ins {
+        Ins::GetVersion => {
+            let version_major = env!("CARGO_PKG_VERSION_MAJOR").parse::<u8>().unwrap();
+            let version_minor = env!("CARGO_PKG_VERSION_MINOR").parse::<u8>().unwrap();
+            let version_patch = env!("CARGO_PKG_VERSION_PATCH").parse::<u8>().unwrap();
+            comm.append([version_major, version_minor, version_patch].as_slice());
+        }
+        Ins::GetPubKey => {
+            let data = comm.get_data()?;
+            if data.len() != 21 {
+                return Err(ErrorCode::BadLen.into());
+            }
+            let raw_path = &data[..20];
+            if !deserialize_path(raw_path, &mut path) {
+                return Err(ErrorCode::BadLen.into());
+            }
+
+            println("raw path");
+            println_slice::<40>(raw_path);
+            let p1 = apdu_header.p1;
+            let p2 = apdu_header.p2;
+            let (pk, hd_index) = derive_pub_key(&mut path, p1, p2)?;
+
+            let need_to_display = data[20] != 0;
+            if need_to_display {
+                review_address(&pk)?;
+            }
+
+            comm.append(pk.as_ref());
+            comm.append(hd_index.to_be_bytes().as_slice());
+        }
+        Ins::SignHash => {
+            let data = comm.get_data()?;
+            if data.len() != 4 * 5 + 32 {
+                return Err(ErrorCode::BadLen.into());
+            }
+            // This check can be removed, but we keep it for double checking
+            if !deserialize_path(&data[..20], &mut path) {
+                return Err(ErrorCode::BadLen.into());
+            }
+
+            match sign_hash_ui(&path, &data[20..]) {
+                Ok((signature_buf, length, _)) => comm.append(&signature_buf[..length as usize]),
+                Err(code) => return Err(code.into()),
+            }
+        }
+        Ins::SignTx => {
+            let data = comm.get_data()?;
```

Comment:
We should call `sign_tx_context.reset();` at the beginning, instead of at the end.

Reply:
We could also reset `tx_reviewer` at the beginning?

Reply:
We cannot reset `SignTxContext` here because a tx may consist of multiple frames. We reset `TxReviewer` when receiving the first frame: https://github.com/alephium/ledger-alephium/blob/d85bcaa245e553b76aabe954919f4faadf9f0dcf/app/src/sign_tx_context.rs#L128

Reply:
But yes, we should call `sign_tx_context.reset()` at the beginning, will fix it.